sample-parent/sample-features/checkpoint-runner/src/main/java/org/talend/sdk/component/sample/feature/checkpoint/CheckpointInput.java

```diff
@@ -0,0 +1,207 @@
+import org.talend.sdk.component.api.configuration.ui.layout.GridLayout;
+import org.talend.sdk.component.api.input.Emitter;
+import org.talend.sdk.component.api.input.Producer;
+import org.talend.sdk.component.api.input.checkpoint.Checkpoint;
+import org.talend.sdk.component.api.input.checkpoint.CheckpointAvailable;
+import org.talend.sdk.component.api.input.checkpoint.CheckpointData;
+import org.talend.sdk.component.api.meta.Documentation;
+
+import lombok.Data;
+
+@Version
+@Icon(value = Icon.IconType.CUSTOM, custom = "icon")
+@Documentation("Checkpoint Input sample processor connector.")
+@Emitter(family = "checkpoint", name = "input")
+public class CheckpointInput implements Serializable {
+
+    private final transient JsonBuilderFactory factory;
+
+    private List<Integer> data;
+
+    private transient ListIterator<Integer> iterator;
+
+    private Integer bookmark;
+
+    private boolean newBookmark;
+
+    private final transient InputConfig configuration;
+
+    public CheckpointInput(final JsonBuilderFactory factory, @Option("configuration") final InputConfig config) {
+        this.factory = factory;
+        this.configuration = config;
+    }
+
+    @PostConstruct
+    public void init() {
+        data = IntStream.range(0, configuration.dataset.maxRecords).boxed().collect(toList());
+        if (configuration.checkpoint == null) {
+            bookmark = 0;
+        } else {
+            if ("finished".equals(configuration.checkpoint.status)) {
+                bookmark = 0;
+            } else {
+                if (configuration.checkpoint.sinceId > configuration.dataset.maxRecords) {
+                    bookmark = 0;
+                } else {
+                    bookmark = configuration.checkpoint.sinceId;
+                }
+            }
+        }
+        if (bookmark == null) {
+            iterator = new ArrayList<Integer>().listIterator();
+        } else {
+            iterator = data.listIterator(bookmark);
+        }
+    }
+
+    @Producer
+    public JsonObject data() {
+        if (iterator == null) {
+            iterator = data.listIterator();
+        }
+        final Integer produced = iterator.hasNext() ? iterator.next() : null;
+        if (produced == null) {
+            configuration.checkpoint.status = "finished";
+            configuration.checkpoint.sinceId = bookmark;
+            return null;
+        }
+        bookmark = produced;
+        configuration.checkpoint.sinceId = produced;
+        configuration.checkpoint.status = "running";
+
+        if (bookmark % 2 == 0) {
+            newBookmark = true;
+        }
+
+        return factory.createObjectBuilder().add("data", configuration.recordPrefix + produced).build();
+    }
+
+    @CheckpointData
+    public Object getCheckpoint() {
+        newBookmark = false;
+        return configuration.checkpoint;
+    }
+
+    @CheckpointAvailable
+    public Boolean isCheckpointReady() {
+        return newBookmark;
```

Comment:
Looks like after init checkpoint is always available. In what scenarios we can call CheckpointAvailable and it will be false?

Reply:
No, it's true whenever we reach the condition and the retrieval wasn't made (@CheckpointData).